### DISCUSSION TITLE
Parallel computing with targets, more questions

### DISCUSSION THREAD

**Discussion by franzbischoff** [ropensci/targets#827] [Help] [answered]
My question may be a broader question, inspired by this one: https://github.com/ropensci/targets/discussions/426, Originally posted by @kkmann

I've experimented with all types of parallel computing for my project, future, clustermq, local and remote.

I'm having a little difficulty on understanding (IMHO, the documentation could be updated with more information) how and where to specify 1) the number of workers, 2) the number of cores, 3) the number of CPUs. I know these concepts may be used interchangeably, but they are not the same in theory.

I understand as:

- **worker**: a `thread`, or `process`, or `session` that starts to build a branch of targets.
- **core**: one "processor unit" inside a CPU.
- **CPU**: a physical "processor unit" that may contain one or more **cores**.

The issues:

1. There are some confusing arguments in clustermq: `clustermq` itself calls each worker a "job", and they can be "multiprocess", "multicore", or remote; still, we can understand that the former can use another CPU, while the "multicore" may be (I can be wrong) constrained to one CPU that have more that one "core" (let's not talk about the `job_size`). `targets` allows us to configure this, but should I configure using `options(n_jobs = 5)` or `tar_resources_clustermq(template = list(num_cores = 5))`? `n_jobs` is the same as `num_cores`? What `tar_make_clustermq(workers = 5)` means in this context?

2. There are some confusing arguments in `future`: `future` itself calls each worker "workers". The "multisession" plan defaults to `availableCores()`, but actually spawns several processes (`availableCores()` returns the number of all the cores from all CPUs). The "multicore" plan tells us in docs that it uses "fork" for splitting, which would make me think of "threads", not "processes". Still, it tells us "separate forked R processes running in the background". `targets` also allows us to configure that, but the current documentation instructs us to use the `future.callr` plan (we can still use other plans). The basic question here resides on using or not the `tar_resources_future(resources = list(n_cores = 5))` (n_cores means cores, threads, processes, workers?). Do we need it for the basic example? What it means exactly when we use `tar_make_future(workers = 5)`?

3. I'm getting confusing behaviors: I'm using nested `tar_map()` from `tarchetypes` ([the code](https://github.com/franzbischoff/false.alarm/blob/592766859adc3731222119d35e6690866bfb1736/scripts/_regime_change.R#L84), not the last version, just to understand the "nested" thing). It creates an object that contains several branches. What I expect is that having no dependencies, "n_workers" branches should be promptly spawned and getting run. I think I only got it when using plan(multicore), but I'm not sure what the appropriate setting is. I had to try several times to figure out how to make it work as expected.

4. How the next target will be processed: Imagine we have this first object with 100 branches and a next object that depends on each of that former object's branches but not all of them. Rephrasing: obj2-branch1 depends on obj1-branch1 and so on. When will obj2-branch1 start? After all obj1 branches are done or as soon as obj1-branch1 is done? How can we make it happen?

Thanks a lot for this package, @wlandau. This is not a complaint (I'm sure you get it), but a way to improve it.

I'm surely available to help where it is possible.

----

More concrete examples:

I've started `tar_make_future(workers = 10)`. As we can see on `tar_watch()`/branches, there are indeed 10 jobs started (there are 229 branches on this object), and they are getting built:

![image](https://user-images.githubusercontent.com/984592/163893864-5561d359-ed01-4e5b-a150-5675144c1d0c.png)

We can check in 'graph' that this is the only object that is being worked on. Its dependency was built, and it is the last object on the pipeline. Strangely we see this on `htop`: 

![image](https://user-images.githubusercontent.com/984592/163893881-2297e514-0aac-4304-bb27-c984532a33c1.png)

There are 20 processes, 10 started when I started the pipeline and 10 were started "now" (I've added the starting time column on next screenshot):

![image](https://user-images.githubusercontent.com/984592/163894997-622b7fb3-15df-4748-b147-3c9c01fc96a5.png)

So, why the jobs are not running? It seems they are waiting and one at a time is running...

:-(

--- 

And! Suddenly more processes start to popping and running at the same time:

![image](https://user-images.githubusercontent.com/984592/163895584-1f02f5c1-168c-4c08-85ca-2ed2b4315e47.png)

That is when another object started to be worked on:

![image](https://user-images.githubusercontent.com/984592/163895646-020e5861-122a-470b-9279-eab0f6792c4e.png)

**Comment by wlandau** [comment, marked as answer]
There's a lot to unpack here.

https://books.ropensci.org/targets/hpc.html mostly explains things in terms of workers: the concurrent external R processes that run the targets. This is the level at which `tar_make_clustermq()` and `tar_make_future()` operate. The `workers` argument in each of these functions is a statement about how many of these processes can exist at a time. At this level, the computing model makes no claim about the underlying hardware. Those R processes could run on a single core, different cores, different CPUs, or different nodes on a computing cluster. 

`targets` itself makes no attempt to configure itself with any hardware implementation in particular. Instead, it leverages packages `clustermq` and `future` because these packages generalize to multiple modes of parallel and distributed computing. `clustermq` and `future` already have really nice interfaces for configuring the parallel computing backend, so `targets` tries to leverage them for setting up `_targets.R`.

When you use `tar_make_clustermq()`, you are also using `clustermq`. Likewise with `tar_make_future()` and `future`. This may appear to come with a set of name conflicts, but it gets easier if you mentally frame it as using 2 or 3 packages instead of just one. The `n_jobs` argument of `clustermq::Q()` corresponds to the `workers` argument of `tar_make_clustermq()`, and the term "workers" in the `future` ecosystem corresponds to the `workers` argument of `tar_make_future()`. Each one is talking about R processes.

*Inside* each R process, you may wish to take advantage of multiple cores or threads. Some R packages like `rstan` and `R2jags` are already multithreaded or have multicore capabilities. If you are running on your local machine, each `targets` worker may get assigned to a different core (or may not), and the things that happen inside each of the targets may also spawn multiple processes across multiple cores. Users of clusters may need to edit the [template file](https://books.ropensci.org/targets/hpc.html#clustermq-remotely) so each `targets` worker has access to multiple cores.

**Reply by wlandau** [reply]
Your pipeline is a bit much for me for a question like this, but I could explore a smaller pipeline to answer a more specific question.

**Reply by franzbischoff** [reply]
Ok, so, what I get from that is (I'll edit this while the answer is getting refined):

- The `workers` argument used on `_targets.yaml` and `tar_make_*()` defines how many `branches` may be running in parallel. We can't make any assumptions at this point if they will use `threads`, `processes`, `cores`, etc.
- The user must specify these next steps on the underlying package (`clustermq` or `future`). `targets` translates the `workers` argument to `n_jobs` for `clustermq` and `workers` for `future`.
- `targets` doesn't create any parallel job by itself. It makes a request to `clustermq`/`future` for a list of `workers` and then deliver the branches for each one.

**Reply by wlandau** [reply]
> The workers argument used on _targets.yaml and tar_make_*() defines how many branches may be running in parallel. We can't make any assumptions at this point if they will use threads, processes, cores, etc.

Exactly. (This applies not only to static and dynamic branches, but also any subset of conditionally independent targets that may or may not make use of any formal branching.)

> The user must specify these next steps on the underlying package (clustermq or future). targets translates the workers argument to n_jobs for clustermq and workers for future.

> First, talking about targets domain. Does targets spawn processes itself or does it make a request to clustermq/future for a list of workers and then deliver the branches for each one?

The latter. `targets` calls `clustermq::workers()` to launch `clustermq` workers, and it calls `future::future()` to launch futures.

**Reply by franzbischoff** [reply]
going further:

- `targets` will send the next available branch to any available `worker` as soon as possible?
- When we use for example, map(variable_with_length_10), `targets` create 10 branches and these 10 branches can be parallelized? If se use a 'batch' of size 5, will be only 2 branches to be parallelized?

**Reply by wlandau** [reply]
> targets will send the next available branch to any available worker as soon as possible?

Yes, as soon as (1) all the upstream dependency targets complete, (2) a worker is available, and (3) no other upcoming targets are ahead in the queue.

> When we use for example, map(variable_with_length_10), targets create 10 branches and these 10 branches can be parallelized?

Yes, those 10 branches can be parallelized because no branch depends on another branch from that group of 10.

> If se use a 'batch' of size 5, will be only 2 branches to be parallelized?

Each batch is a dynamic branch target. So with 10 tasks partitioned into batches of size 5, each target will do 5 tasks, so there are 2 targets. Those targets do not depend on each other, so they can can run in parallel.

Parallelism is dynamic. It depends on the structure of the dependency graph, how many workers are available, and the variability in how long it takes each target to run. There is a large variety of possible dependency graph shapes, and dynamic branching changes the shape of the graph while the pipeline is running. The scheduler in `targets` takes the overall high-level view of this graph shape without paying specific attention to where that shape came from, e.g. any specific calls to `pattern = map(variable)`.

**Reply by franzbischoff** [reply]
soon more info here. Been playing with futures etc :-) I managed to overload 40CPU's + 190GB mem lol

**Reply by wlandau** [reply]
`tar_target()` and `tar_option_set()` have arguments `storage`, `retrieval`, `memory`, and `garbage_collection`. For heavy data workflows that risk maxing out memory, you might consider setting these values to `"worker"`, `"worker"`, `"transient"`, and `TRUE`, respectively.
